File: 70. Climbing Stairs.java
/**
 * https://leetcode.com/problems/climbing-stairs/
 */

class Solution {
    // O(n) time complexity
    // O(n) space complexity
    public int climbStairs(int n) {
        if (n == 1) return 1;
        int[] m = new int[n];
        m[0] = 1;
        m[1] = 2;

        for (int i = 2; i < n; i++) {
            m[i] = m[i - 1] + m[i - 2];
        }
        return m[n - 1];

    }
}

File: Kotlin/70. Climbing Stairs.kt
/**
 * https://leetcode.com/problems/climbing-stairs/
 */

class Solution {
    // O(n) time complexity
    // O(n) space complexity
    fun climbStairs(n: Int): Int {
        if (n == 1) return 1
        val arr = IntArray(n)
        arr[0] = 1
        arr[1] = 2
        for (i in 2 until n)
            arr[i] = arr[i - 1] + arr[i - 2]

        return arr[n - 1]
    }
}